### PR TITLE
Fix columns order after Block::sortColumns()

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix columns order after Block::sortColumns() (also add a test that shows that it affects some real use case - Buffer engine)

Fixes: #9950

### CODE REVIEW

File: src/Core/Block.cpp [hunk 1]
```diff
@@ -403,7 +404,12 @@ Block Block::sortColumns() const
 {
     Block sorted_block;
 
+    /// std::unordered_map cannot be used to guarantee the sort order
```
Comment (alexey-milovidov) [inline src/Core/Block.cpp:407]: Absolutely right.
It looks like someone was replaced std::map to std::unordered_map without paying attention.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Note that filling std::vector and calling std::sort is more efficient than creating std::map (both time and memory). And also you can sort iterators or even value_type pointers.

**Comment by alexey-milovidov** [comment]
Was broken here: https://github.com/ClickHouse/ClickHouse/pull/9950

**Comment by azat** [comment]
>Was broken here: #9950

Didn't even though that there was std::map (hence didn't even looked into the `git blame`)

>Note that filling std::vector and calling std::sort is more efficient than creating std::map (both time and memory). And also you can sort iterators or even value_type pointers.

Indeed, nice catch, will fix it shortly

File: src/Core/Block.cpp (hunk 1 above)
Comment (azat) [reply src/Core/Block.cpp:407]: >without paying attention.

Let's blame lack of test :)

**Comment by azat** [comment]
>you can sort iterators

Done

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**